tools/build_langserver/langserver/completion.go

```diff
@@ -0,0 +1,263 @@
+			supportSnippet, pos, partial)
+	} else if LooksLikeCONFIGAttr(lineContent) {
+		// TODO(bnm): Perhaps this can be extracted to itemsFromProperty
+		var completionList []*lsp.CompletionItem
+		for tag, field := range analyzer.State.Config.TagsToFields() {
+			if !strings.Contains(tag, strings.ToUpper(partial)) {
+				continue
+			}
+			TERange := getTERange(pos, partial)
+			item := getCompletionItem(lsp.Property, tag, field.Tag.Get("help"),
+				nil, supportSnippet, TERange)
+
+			completionList = append(completionList, item)
+		}
+		return completionList
+	}
+	return nil
+}
+
+// partial: the string partial of the Attribute
+func itemsFromMethods(attributes []*RuleDef, supportSnippet bool,
+	pos lsp.Position, partial string) []*lsp.CompletionItem {
+
+	var completionList []*lsp.CompletionItem
+	for _, attr := range attributes {
+		// Continue if the the name is not part of partial
+		if !strings.Contains(attr.Name, partial) {
+			continue
+		}
+		item := itemFromRuleDef(attr, supportSnippet, pos, partial)
+		completionList = append(completionList, item)
+	}
+
+	return completionList
+}
+
+// Gets all completion items from function or method calls
+func itemFromRuleDef(ruleDef *RuleDef, supportSnippet bool,
+	pos lsp.Position, partial string) *lsp.CompletionItem {
+
+	// Get the first line of docString as lsp.CompletionItem.detail
+	docStringList := strings.Split(ruleDef.Docstring, "\n")
+	var detail string
+	if len(docStringList) > 0 {
+		detail = docStringList[0]
+	}
+
+	TERange := getTERange(pos, partial)
+	return getCompletionItem(lsp.Function, ruleDef.Name, detail, ruleDef, supportSnippet, TERange)
+}
+
+func getCompletionItem(kind lsp.CompletionItemKind, name string,
+	detail string, ruleDef *RuleDef, supportSnippet bool, TERange lsp.Range) *lsp.CompletionItem {
+
+	var format lsp.InsertTextFormat
+	var text string
+	if kind == lsp.Function && supportSnippet && ruleDef != nil {
+		// Get the snippet for completion
+		snippet := name + "("
+		if len(ruleDef.ArgMap) > 0 {
+			for _, arg := range ruleDef.Arguments {
+				if arg.Name != "self" && ruleDef.ArgMap[arg.Name].required == true {
+					snippet += arg.Name
+				}
+			}
+		}
+		snippet += ")"
+		format, text = lsp.ITFSnippet, snippet
+	} else {
+		format, text = lsp.ITFPlainText, name
+	}
+
+	return &lsp.CompletionItem{
+		Label:            name,
+		Kind:             kind,
+		Detail:           detail,
+		InsertTextFormat: format,
+
+		// InsertText is deprecated in favour of TextEdit, but added here for legacy client support
+		InsertText: text,
+		TextEdit: &lsp.TextEdit{
+			Range:   TERange,
+			NewText: text,
+		},
+	}
+}
+
+func getTERange(pos lsp.Position, partial string) lsp.Range {
+	return lsp.Range{
+		Start: lsp.Position{Line: pos.Line, Character: pos.Character - len(partial)},
+		End:   pos,
+	}
+}
+
+func isVisible(buildDef *BuildDef, currentPkg string) bool {
+	for _, i := range buildDef.Visibility {
+		if i == "PUBLIC" || strings.HasPrefix(i, currentPkg) {
```

Comment:
This isn't quite correct... you should be using the `Includes` function of `BuildLabel`